### ISSUE TITLE
improved cert validation, IETF RFC 9618

### ISSUE THREAD

**Issue by richsalz** [quictls/quictls#61]
Perhaps we can get @davidben or Bob Beck to point us to the code in boring?

See RFC 9618

This document updates RFC 5280 to replace the algorithm for X.509
policy validation with an equivalent, more efficient algorithm. The
original algorithm built a structure that scaled exponentially in the
worst case, leaving implementations vulnerable to denial-of-service
attacks.

**Comment by botovq** [comment]
It's here:

https://github.com/google/boringssl/blob/master/crypto/x509/policy.c

We ported this without too much trouble, but we may have been helped by
already having purged the old public API interface of the policy code.

One bit where one has to be really careful is the size_t changes
in BoringSSL's STACK_OF() API.

**Comment by wbl** [comment]
We could also drop policies entirely: I don't know of many uses, there aren't any in the WebPKI, and its an optional part of 5280.

**Comment by bob-beck** [comment]
We only kept it in Boring because certain US government agencies use it extensively in private pki that needs to be supported. Beyond that, and RPKI, I don't know of many uses.  Having said that FedRamp stuff may be a reason why you want to keep an implementation, at which point it's probably worth fixing.

**Comment by davidben** [comment]
I don't think that was actually the reason, though yes the FPKI is one of the main uses of all these features. EV certificates are also indicated with policies, though they use only a tiny tiny subset of the features.

The actual concern at the time wasn't very specific. The verifier derived from OpenSSL originally inherited OpenSSL's bug (upstream decided not to fix it) where, by default, it doesn't evaluate policies, but still accepts critical policy extensions. This means that you might accept a chain that you shouldn't have by policy validation, if the chain had a critical policy constraints that asserted requireExplicitPolicy.

Honestly, what's the point of constraining an intermediate that way, I do not know, but still, I think any correct X.509 implementation is obligated to either implement policy validation, or reject all critical policy-related extensions that might cause a chain to be rejected at path validation. Given the pre-existing behavior, that some folks apparently use it (misguided as it is), and the need to do something about the broken IETF specification, it seemed better to just fix the algorithm and demonstrate it was possible.
